p8e-api/src/main/kotlin/io/provenance/engine/service/ChaincodeInvokeService.kt

```diff
@@ -40,242 +55,268 @@ class ChaincodeInvokeService(
     private val objectMapper = ObjectMapper().configureProvenance()
     private val indexRegex = "^.*message index: (\\d+).*$".toRegex()
 
-    private val queue = ConcurrentHashMap<UUID, BlockchainTransaction>()
+    // private val queue = ConcurrentHashMap<UUID, BlockchainTransaction>()
+
+    // thread safe queue because this spans the worker thread and the enqueue
+    private val queue = LinkedBlockingQueue<ContractRequestWrapper>(1_000)
+
+    // non-thread safe data structures that are only used within the worker thread
+    private val batch = HashSet<ContractRequestWrapper>()
+    private val blockScopeIds = HashSet<String>()
+    private val priorityScopeBacklog = HashMap<String, LinkedList<ContractRequestWrapper>>()
 
     init {
         thread(isDaemon = true, name = "bc-tx-batch") {
             memorializeBatchTx()
         }
     }
 
-    private data class BlockchainTransaction(val txPreparer: TxPreparer, val future: CompletableFuture<ContractTxResult>, val executionUuid: UUID, var attempts: Int = 0)
-
     fun memorializeBatchTx() {
         log.info("Starting bc-tx-batch thread")
 
-        var emptyIterations = 0
-
         while(true) {
-            Thread.sleep(chaincodeProperties.emptyIterationBackoffMS.toLong())
+            // attempt to load the batch with scopes that were previously passed on due to not
+                // wanting to send the same scope in the same block
+            while (batch.size < chaincodeProperties.txBatchSize) {
+                // filter the backlog for scopes that aren't in the block and then pick a random
+                    // scope and insert it into the block
+                priorityScopeBacklog.filterKeys { !blockScopeIds.contains(it) }
+                    .keys
+                    .toList()
+                    .takeIf { it.isNotEmpty() }
+                    ?.let { keyList ->
+                        val key = keyList.get(Random.nextInt(keyList.size))
+                        val messages = priorityScopeBacklog.getValue(key)
+                        val message = messages.pop()
+
+                        if (messages.isEmpty()) {
+                            log.debug("removing scope $key from backlog as its list is now empty")
+
+                            priorityScopeBacklog.remove(key)
+                        }
 
-            val batch = mutableMapOf<UUID, BlockchainTransaction>()
+                        blockScopeIds.add(message.request.scopeId)
```

Comment:
probably should have a matching debug message to the one in the queue.poll block below (or move the message and blockScopeIds/batch.add to a separate function)